Collection deposits search

jmartin-sul opened the pull request:
closes #1021 
<img width="1165" alt="Screenshot 2025-06-13 at 3 23 04 PM" src="https://github.com/user-attachments/assets/abc623fe-bd51-4313-9581-1a13fc006183" />
<img width="1126" alt="Screenshot 2025-06-13 at 3 23 16 PM" src="https://github.com/user-attachments/assets/d5b5c151-9812-4a8d-94e2-3a7017e7c9fa" />
<img width="1134" alt="Screenshot 2025-06-13 at 3 23 24 PM" src="https://github.com/user-attachments/assets/0e9582fb-d2c2-42f8-828d-a24e58fa1fa6" />
<img width="1152" alt="Screenshot 2025-06-13 at 3 23 35 PM" src="https://github.com/user-attachments/assets/d2c6f1f7-36e4-422c-a683-a8e091dc67ae" />
<img width="1136" alt="Screenshot 2025-06-13 at 3 23 45 PM" src="https://github.com/user-attachments/assets/03a7ec6b-d81f-4c02-957c-6b6daa0efdd0" />
<img width="1140" alt="Screenshot 2025-06-13 at 3 23 59 PM" src="https://github.com/user-attachments/assets/1b7134e2-95f1-4c6e-8824-2f523c21f2eb" />

jmartin-sul commented on app/controllers/collections_controller.rb:
```diff
@@ -79,10 +79,26 @@ def wait
+      ).or(
+        User.where('first_name ILIKE ?', "%#{@search_term}%")
+      )
+    end
```
happy to factor out the updated query logic into a helper method, if people would prefer that to disabling the rubocop complaint.  i found this readable, but could've gone either way.

jmartin-sul commented:
leaving in draft till i can discuss the styling on monday, since this sorta deviates from the mockups.

jmartin-sul commented on spec/system/show_collection_spec.rb:
```diff
@@ -209,6 +209,54 @@
         end
       end
 
+      within('div#deposits-pane') do
```
ended up leaving a variation with results and a variation without for each user type, and moved everything else to request specs.  happy to pare down even further if you want.

justinlittman commented on app/controllers/collections_controller.rb:
```diff
@@ -7,6 +7,7 @@ class CollectionsController < ApplicationController
   before_action :set_collection_form_from_cocina, only: %i[show edit]
   before_action :set_status, only: %i[show edit]
   before_action :set_presenter, only: %i[show edit]
+  before_action :set_search_term, only: %i[works]
```
If this is only used here in one method then would be clearer if inline in the method.

jmartin-sul replied:
was on the fence about that one, had it that way at first, flipping it back...

justinlittman commented on app/controllers/collections_controller.rb:
```diff
@@ -100,6 +101,27 @@ def history
 
+  # @return [ActiveRecord::Relation] the works to display for the collection, filtered by search term if applicable
+  def collection_works_result
+    collection_works_result = @collection.works
```
Slight suggestion:
```
collection_works_result = @collection.works

return collection_work_result if @search_term.blank?

collection_works_result.joins(
 ...
```

jmartin-sul replied:
ended up tightening it up a bit more than that, and got rid of the `collection_works_result` temp variable, since it doesn't really get built up anymore after this refactoring.  if there's more conditional filtering or sorting to be added later, someone can add a temp var for building the `ActiveRecord::Relation` at that point.  (i'd put in the temp var in when the query building was directly in the controller method, primarily because i thought that made it slightly more readable there -- now that there's a separate helper method, the temp var feels like it adds less to me)

justinlittman reviewed (approved):
Approve pending consideration of comments.

jmartin-sul commented:
ok, updated with the requested tweaks.  here's the diff from right before i rebased my changes on `origin/main`:

```diff
% git diff origin/collection-deposits-search HEAD      collection-deposits-search
diff --git a/app/controllers/collections_controller.rb b/app/controllers/collections_controller.rb
index 4fa922b5..47f982ac 100644
--- a/app/controllers/collections_controller.rb
+++ b/app/controllers/collections_controller.rb
@@ -7,7 +7,6 @@ class CollectionsController < ApplicationController
   before_action :set_collection_form_from_cocina, only: %i[show edit]
   before_action :set_status, only: %i[show edit]
   before_action :set_presenter, only: %i[show edit]
-  before_action :set_search_term, only: %i[works]
 
   def show
     authorize! @collection
@@ -83,6 +82,7 @@ class CollectionsController < ApplicationController
   def works
     authorize! @collection
 
+    @search_term = params[:q]
     @works = authorized_scope(collection_works_result, as: :collection, scope_options: { collection: @collection })
              .order(updated_at: :desc).page(params[:page])
     @work_statuses = Sdr::Repository.statuses(
@@ -103,23 +103,19 @@ class CollectionsController < ApplicationController
 
   # @return [ActiveRecord::Relation] the works to display for the collection, filtered by search term if applicable
   def collection_works_result
-    collection_works_result = @collection.works
-
-    if @search_term.present?
-      collection_works_result = collection_works_result.joins(
-        :user
-      ).where('title ILIKE ?', "%#{@search_term}%").or(
-        Work.where('druid ILIKE ?', "%#{@search_term}%")
-      ).or(
-        User.where('email_address ILIKE ?', "%#{@search_term}%")
-      ).or(
-        User.where('name ILIKE ?', "%#{@search_term}%")
-      ).or(
-        User.where('first_name ILIKE ?', "%#{@search_term}%")
-      )
-    end
-
-    collection_works_result
+    return @collection.works if @search_term.blank?
+
+    @collection.works.joins(
+      :user
+    ).where('title ILIKE ?', "%#{@search_term}%").or(
+      Work.where('druid ILIKE ?', "%#{@search_term}%")
+    ).or(
+      User.where('email_address ILIKE ?', "%#{@search_term}%")
+    ).or(
+      User.where('name ILIKE ?', "%#{@search_term}%")
+    ).or(
+      User.where('first_name ILIKE ?', "%#{@search_term}%")
+    )
   end
 
   def collection_params
@@ -134,10 +130,6 @@ class CollectionsController < ApplicationController
     @collection = Collection.find_by!(druid: params[:druid])
   end
 
-  def set_search_term
-    @search_term = params[:q]
-  end
-
   def check_deposit_registering_or_updating
     redirect_to wait_collections_path(@collection.id) if @collection.deposit_registering_or_updating?
   end
```

will merge when green, since approved.
